### ISSUE TITLE
ipfs dht provide is not parallelized

### ISSUE THREAD

**Issue by kyledrake** [ipfs/kubo#4863]
Enhancement:

The `ipfs dht provide CID ...` command is not currently parallelized. This means that if a larger list of records to provide is given, it runs pretty slow because it has to run the provides sequentially.

A better approach here would be to batch X number of provides at a time in parallel when a list of many cids is provided, so the end user doesn't have to parallelize on their end.

This is a lower priority improvement. Workaround is to do the parallelizing upstream, or use something like gnu parallel.

**Comment by magik6k** [comment]
This will be really easy to do once https://github.com/ipfs/go-ipfs/pull/4333 is done

**Comment by magik6k** [comment]
cc @michaelavila

**Comment by michaelavila** [comment]
I'll take a look at this with the rest of the provider work. Thanks for pointing it out.

**Comment by michaelavila** [comment]
Just an update:

After a brief discussion [here](https://github.com/ipfs/go-ipfs/pull/6396) it was decided to leave the `ipfs dht provide <hash>` command alone as its a useful debugging mechanism. We are also in the process of a building a more robust provider system that has facilities for responsibly providing many cids at once. The command you will likely want to use in the future is `ipfs provider provide <hash>`. Although this particular subcommand has not been built yet, it's not far off.

I'm going to leave this issue open until the suggested solution is available to use. Future issues filed about parallelizing `ipfs dht provide` should reference here.

**Comment by guillaumemichel** [comment]
Implemented in https://github.com/ipfs/boxo/pull/907
